### ISSUE TITLE
several segfaults new in bleedperl

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2322]
Migrated from [rt.perl.org#3646](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3646) (status was 'resolved')

Searchable as RT3646$

**Comment by p5pRT** [comment]
### From jfriedl@yahoo-inc.com

#### Created by jfriedl@yahoo-inc.com



I've been busy today \-\- what follows are reports on seven mostly\-different
situations where perl will either segfault or panic\. Some of these are new
in bleedperl \(as of Aug 6\)\.

They're all related in that they all use \(??\{\.\.\.\}\)\.




Problem \#1&#8203;: the following program segfaults with bleedperl\, but not 5\.6\.0&#8203;:

&nbsp;    use strict;
&nbsp;    use re 'eval';

&nbsp;    $\_ = "1";
&nbsp;    our $foo = qr\{ \(\\d\+\)
&nbsp;            \(?\{ warn "foo" \}\)
&nbsp;            \(?\{ warn $undefined \}\) \}x;

&nbsp;    m\{ \(??\{ $foo \}\) \}x;


With 5\.6\.0&#8203;:
&nbsp;    foo at \(re\_eval 1\) line 1\.
&nbsp;    Use of uninitialized value in warn at \(re\_eval 2\) line 1\.
&nbsp;    Warning&#8203;: something's wrong at \(re\_eval 2\) line 1\.

With bleedperl&#8203;:

&nbsp;    foo at \(re\_eval 1\) line 1\.
&nbsp;    Segmentation fault

Here's the backtrace

&nbsp;    jfriedl@&#8203;fummy> gdb \./perl
&nbsp;    GNU gdb 4\.18
&nbsp;    Copyright 1998 Free Software Foundation\, Inc\.
&nbsp;    \(gdb\) run /tmp/p
&nbsp;    Starting program&#8203;: /home/jfriedl/src/bleadperl/\./perl /tmp/p
&nbsp;    foo at \(re\_eval 1\) line 1\.

&nbsp;    Program received signal SIGSEGV\, Segmentation fault\.
&nbsp;    malloc \(nbytes=1\) at malloc\.c&#8203;:1067
&nbsp;    1067            nextf\[bucket\] = p\->ov\_next;
&nbsp;    \(gdb\) where
&nbsp;    \#0  malloc \(nbytes=1\) at malloc\.c&#8203;:1067
&nbsp;    \#1  0x809e006 in Perl\_sv\_grow \(sv=0x80f5964\, newlen=1\) at sv\.c&#8203;:1246
&nbsp;    \#2  0x80a0ee0 in Perl\_sv\_setpvn \(sv=0x80f5964\, ptr=0x80e3338 ""\, len=0\) at sv\.c&#8203;:3006
&nbsp;    \#3  0x80a3679 in Perl\_newSVpvn \(s=0x80e3338 ""\, len=0\) at sv\.c&#8203;:4901
&nbsp;    \#4  0x808c95b in S\_mess\_alloc \(\) at util\.c&#8203;:1374
&nbsp;    \#5  0x808ca91 in Perl\_vmess \(pat=0x80e7ea0 "Use of uninitialized value%s%s"\, args=0xbfffeccc\)
&nbsp;    at util\.c&#8203;:1450
&nbsp;    \#6  0x808d679 in Perl\_vwarner \(err=41\, pat=0x80e7ea0 "Use of uninitialized value%s%s"\, args=0xbfffeccc\)
&nbsp;    at util\.c&#8203;:1819
&nbsp;    \#7  0x808d64e in Perl\_warner \(err=41\, pat=0x80e7ea0 "Use of uninitialized value%s%s"\) at util\.c&#8203;:1804
&nbsp;    \#8  0x809d126 in Perl\_report\_uninit \(\) at sv\.c&#8203;:304
&nbsp;    \#9  0x809fc0a in Perl\_sv\_2pv \(sv=0x810a338\, lp=0xbfffed94\) at sv\.c&#8203;:2273
&nbsp;    \#10 0x80bdeca in Perl\_pp\_warn \(\) at pp\_sys\.c&#8203;:431
&nbsp;    \#11 0x8096e4d in Perl\_runops\_standard \(\) at run\.c&#8203;:27
&nbsp;    \#12 0x80d1d69 in S\_regmatch \(prog=0x8100af8\) at regexec\.c&#8203;:2366
&nbsp;    \#13 0x80d3668 in S\_regmatch \(prog=0x8100ae8\) at regexec\.c&#8203;:3117
&nbsp;    \#14 0x80d1f69 in S\_regmatch \(prog=0x80f615c\) at regexec\.c&#8203;:2435
&nbsp;    \#15 0x80d0669 in S\_regtry \(prog=0x80f6120\, startpos=0x80f5020 "1"\) at regexec\.c&#8203;:1774
&nbsp;    \#16 0x80d02ff in Perl\_regexec\_flags \(prog=0x80f6120\, stringarg=0x80f5020 "1"\, strend=0x80f5021 ""\, 
&nbsp;    strbeg=0x80f5020 "1"\, minend=0\, sv=0x80f5910\, data=0x0\, flags=2\) at regexec\.c&#8203;:1630
&nbsp;    \#17 0x80992de in Perl\_pp\_match \(\) at pp\_hot\.c&#8203;:1082
&nbsp;    \#18 0x8096e4d in Perl\_runops\_standard \(\) at run\.c&#8203;:27
&nbsp;    \#19 0x805c12e in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1447
&nbsp;    \#20 0x805bea8 in perl\_run \(my\_perl=0x80f5004\) at perl\.c&#8203;:1367
&nbsp;    \#21 0x8059c83 in main \(argc=2\, argv=0xbffff354\, env=0xbffff360\) at perlmain\.c&#8203;:52
&nbsp;    \(gdb\) 

My bleedperl is compiled with Perl's malloc\. I had another problem earlier
in which I was getting segfaults in libc\, and although I spent a few hours
with gdb\, I couldn't even track down where from within Perl the errant libc
function was being called\. When I recompiled with Perl's malloc\, the
problem went away\. I don't know if it's related to this\, though\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Problem\#2&#8203;:
&nbsp;  As I noted before\, program execution results in&#8203;:
&nbsp;      foo at \(re\_eval 1\) line 1\.
&nbsp;      Segmentation fault
&nbsp;  However\, if you comment out the "use re 'eval'" line\, you then get
&nbsp;       at \(re\_eval 1\) line 1\.
&nbsp;      Segmentation fault
&nbsp;  which seems odd\.  Again\, this oddness does not occur with 5\.6\.0\.
&nbsp;  Using "no re 'eval'" doesn't make the oddness go away\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Problem\#3&#8203;:
&nbsp;  In trying to make a t/op test for problem\#2\, I ran into yet another
&nbsp;  problem\. The program&#8203;:

&nbsp;      eval \{
&nbsp;      $\_ = "1";
&nbsp;      our $foo = qr\{ \(\\d\+\)
&nbsp;             \(?\{ die "foo" \}\)
&nbsp;              \(?\{ $undefined \}\) \}x;
&nbsp;      m\{ \(??\{ $foo \}\) \}x;
&nbsp;      \};

&nbsp;      print "result is \[$@&#8203;\]\\n";

&nbsp;  Generates&#8203;:
&nbsp;   Modification of a read\-only value attempted at t/op/regex\-oddness\.t line 11\.


&nbsp;  I don't think there's anything read\-only about $foo or $\_\, the two variables
&nbsp;  that are modified\.

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Problem\#4&#8203;:

&nbsp;  If you then add a print before the result\-print line\,

&nbsp;       eval \{
&nbsp;       $\_ = "1";
&nbsp;       our $foo = qr\{ \(\\d\+\)
&nbsp;              \(?\{ die "foo" \}\)
&nbsp;               \(?\{ $undefined \}\) \}x;
&nbsp;       m\{ \(??\{ $foo \}\) \}x;
&nbsp;       \};

&nbsp;       print "before\\n"
&nbsp;       print "result is \[$@&#8203;\]\\n";

&nbsp;  it then segfaults&#8203;:

&nbsp;    jfriedl@&#8203;fummy> gdb \./perl
&nbsp;    \(gdb\) run t/op/regex\-oddness\.t
&nbsp;    Starting program&#8203;: /home/jfriedl/src/bleadperl/\./perl t/op/regex\-oddness\.t
&nbsp;    1\.\.2
&nbsp;    before

&nbsp;    Program received signal SIGSEGV\, Segmentation fault\.
&nbsp;    0x808fa0c in Perl\_mg\_get \(sv=0x80fe050\) at mg\.c&#8203;:111
&nbsp;    111             if \(\!\(mg\->mg\_flags & MGf\_GSKIP\) && vtbl && vtbl\->svt\_get\) \{
&nbsp;    \(gdb\) where
&nbsp;    \#0  0x808fa0c in Perl\_mg\_get \(sv=0x80fe050\) at mg\.c&#8203;:111
&nbsp;    \#1  0x80976be in Perl\_pp\_concat \(\) at pp\_hot\.c&#8203;:210
&nbsp;    \#2  0x8096e4d in Perl\_runops\_standard \(\) at run\.c&#8203;:27
&nbsp;    \#3  0x805c12e in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1447
&nbsp;    \#4  0x805bea8 in perl\_run \(my\_perl=0x80f5004\) at perl\.c&#8203;:1367
&nbsp;    \#5  0x8059c83 in main \(argc=2\, argv=0xbffff344\, env=0xbffff350\) at perlmain\.c&#8203;:52
&nbsp;    \(gdb\)


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-


Problem\#5&#8203;:

&nbsp;  If you take the first program above and inline the $foo regex directly into
&nbsp;  the main regex\, as with&#8203;:

&nbsp;    use strict;
&nbsp;    use re 'eval';

&nbsp;    $\_ = "1";

&nbsp;    m\{ \(??\{ qr\{\(\\d\+\)
&nbsp;            \(?\{ warn "foo" \}\)
&nbsp;            \(?\{ warn $undefined \}\)\} \}\) \}x;

&nbsp;  you then get
&nbsp;      panic&#8203;: unknown regstclass 0 at /tmp/p line 9\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Problem\#6&#8203;:

 Making the program even smaller

&nbsp;     use re 'eval';
&nbsp;     m\{ \(??\{ qr\{\(\\d\+\)\} \}\) \}x;

 and you get a segfault \(gdb dump below\)\.
 Oddly\, with 5\.6\.0\, you get the line
&nbsp;      80fe468 176
 before the segfault\. With bleedperl\, you don't get it\.

 Here's the gdb dump&#8203;: 

&nbsp;    jfriedl@&#8203;fummy> gdb \./perl
&nbsp;    \(gdb\) run /tmp/p
&nbsp;    Starting program&#8203;: /home/jfriedl/src/bleadperl/\./perl /tmp/p

&nbsp;    Program received signal SIGSEGV\, Segmentation fault\.
&nbsp;    0x8083d06 in S\_study\_chunk \(scanp=0xbfffef88\, deltap=0xbfffef84\, last=0x80f6178\, data=0xbffff128\, 
&nbsp;    flags=3072\) at regcomp\.c&#8203;:466
&nbsp;    466                        && off \+ noff \< max\)
&nbsp;    \(gdb\) where
&nbsp;    \#0  0x8083d06 in S\_study\_chunk \(scanp=0xbfffef88\, deltap=0xbfffef84\, last=0x80f6178\, data=0xbffff128\, 
&nbsp;    flags=3072\) at regcomp\.c&#8203;:466
&nbsp;    \#1  0x8084871 in S\_study\_chunk \(scanp=0xbffff0cc\, deltap=0xbffff0d0\, last=0x80f6178\, data=0xbffff128\, 
&nbsp;    flags=3072\) at regcomp\.c&#8203;:732
&nbsp;    \#2  0x80866f9 in Perl\_pregcomp \(exp=0x8101b98 " \(??\{ qr\{\(\\\\d\+\)\} \}\) "\, xend=0x8101bab ""\, pm=0x81012c0\)
&nbsp;    at regcomp\.c&#8203;:1537
&nbsp;    \#3  0x807c5d3 in Perl\_pmruntime \(o=0x81012c0\, expr=0x80f7aa0\, repl=0x0\) at op\.c&#8203;:2901
&nbsp;    \#4  0x8077cb1 in Perl\_yyparse \(\) at perly\.y&#8203;:650
&nbsp;    \#5  0x805bccb in S\_parse\_body \(env=0x0\, xsinit=0x8059cb0 \<xs\_init>\) at perl\.c&#8203;:1290
&nbsp;    \#6  0x805b412 in perl\_parse \(my\_perl=0x80f5004\, xsinit=0x8059cb0 \<xs\_init>\, argc=2\, argv=0xbffff354\, 
&nbsp;    env=0x0\) at perl\.c&#8203;:875
&nbsp;    \#7  0x8059c6f in main \(argc=2\, argv=0xbffff354\, env=0xbffff360\) at perlmain\.c&#8203;:50
&nbsp;    \(gdb\) 


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-


Problem \#7&#8203;: then\, removing the "use re 'eval'" from that small one\, you
get the panic again&#8203;:

&nbsp;    jfriedl@&#8203;fummy> \./perl \-we 'm\{ \(??\{ qr\{\(\\d\+\)\} \}\) \}x;'
&nbsp;    Use of uninitialized value in pattern match \(m//\) at \-e line 1\.
&nbsp;    panic&#8203;: unknown regstclass 0 at \-e line 1\.




I don't know how related these all are\.
&nbsp;    Jeffrey

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jfriedl at Sat Aug  5 23:33:33 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.16, archname=i686-linux
    uname='linux fummy.dsl.yahoo.com 2.2.16 #6 smp sun jul 23 11:26:16 pdt 2000 i686 unknown '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -g', gccversion=pgcc-2.91.66 19990314 (egcs-1.1.2 release), gccosandvers=
    cppflags='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    devel-6489


@INC for perl v5.6.0:
    /home/jfriedl/lib/perl
    /home/jfriedl/lib/perl/yahoo
    /usr/test/lib/perl5/5.6.0/i686-linux
    /usr/test/lib/perl5/5.6.0
    /usr/test/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/test/lib/perl5/site_perl/5.6.0
    /usr/test/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/jfriedl
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/local/pgsql/lib:/home/jfriedl/src/rvplayer5.0
    LOGDIR (unset)
    PATH=/home/jfriedl/bin:/home/jfriedl/common/bin:/usr/local/gcc-2.95.2/bin:.:/usr/local/pgsql/bin:/usr/local/bin:/usr/X11R6/bin:/bin:/usr/bin:/usr/sbin:/sbin:/home/jfriedl/src/rvplayer5.0
    PERLLIB=/home/jfriedl/lib/perl:/home/jfriedl/lib/perl/yahoo
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @jhi

> I've been busy today \-\- what follows are reports on seven mostly\-different
> situations where perl will either segfault or panic\. Some of these are new
> in bleedperl \(as of Aug 6\)\.
> 
> They're all related in that they all use \(??\{\.\.\.\}\)\.

Sigh\.  The \(?? truly is experimental\.

**Comment by p5pRT** [comment]
### From @vanstyn

In \<20000807112700\.A2773@&#8203;chaos\.wustl\.edu>\, Jarkko Hietaniemi writes&#8203;:
:> I've been busy today \-\- what follows are reports on seven mostly\-different
:> situations where perl will either segfault or panic\. Some of these are new
:> in bleedperl \(as of Aug 6\)\.
:> 
:> They're all related in that they all use \(??\{\.\.\.\}\)\.
:
:Sigh\.  The \(?? truly is experimental\.

I have all of Jeffrey's recent reports filed for checking\, but I
suspect that most or all of the \(??\{\.\.\.\}\) problems will require that
the engine become reentrant\, and possibly that we also replace the
way we jump out of scopes through JMPENV\. ISTR Sarathy had some ideas
about the latter\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

With problem \#6 on NT with 5\.6\.0\, it looks like it all starts with an
invalid pointer read&#8203;:
&nbsp;        S\_nextchar     \[regcomp\.c&#8203;:3575\]
&nbsp;        
&nbsp;            for \(;;\) \{
&nbsp;            if \(\*PL\_regcomp\_parse == '\(' && PL\_regcomp\_parse\[1\] == '?' &&
&nbsp;     =>         PL\_regcomp\_parse\[2\] == '\#'\) \{
&nbsp;                while \(\*PL\_regcomp\_parse && \*PL\_regcomp\_parse \!= '\)'\)
&nbsp;                PL\_regcomp\_parse\+\+;
&nbsp;                PL\_regcomp\_parse\+\+;
&nbsp;        S\_reg          \[regcomp\.c&#8203;:1791\]
&nbsp;                    FAIL\("Eval\-group in insecure regular expression"\);
&nbsp;                \}
&nbsp;        
&nbsp;     =>         nextchar\(\);
&nbsp;                if \(logical\) \{
&nbsp;                    ret = reg\_node\(LOGICAL\);
&nbsp;                    if \(\!SIZE\_ONLY\)
&nbsp;        S\_regatom      \[regcomp\.c&#8203;:2297\]
&nbsp;            break;
&nbsp;            case '\('&#8203;:
&nbsp;            nextchar\(\);
&nbsp;     =>     ret = reg\(1\, &flags\);
&nbsp;            if \(ret == NULL\) \{
&nbsp;                if \(flags & TRYAGAIN\)
&nbsp;                    goto tryagain;
&nbsp;        S\_regpiece     \[regcomp\.c&#8203;:2096\]
&nbsp;            I32 min;
&nbsp;            I32 max = REG\_INFTY;
&nbsp;        
&nbsp;     =>     ret = regatom\(&flags\);
&nbsp;            if \(ret == NULL\) \{
&nbsp;            if \(flags & TRYAGAIN\)
&nbsp;                \*flagp |= TRYAGAIN;
&nbsp;        S\_regbranch    \[regcomp\.c&#8203;:2044\]
&nbsp;            nextchar\(\);
&nbsp;            while \(PL\_regcomp\_parse \< PL\_regxend && \*PL\_regcomp\_parse \!= '|'
&& \*PL\_regcomp\_parse \!= '\)'\) \{
&nbsp;            flags &= ~TRYAGAIN;
&nbsp;     =>     latest = regpiece\(&flags\);
&nbsp;            if \(latest == NULL\) \{
&nbsp;                if \(flags & TRYAGAIN\)
&nbsp;                continue;
&nbsp;        S\_reg          \[regcomp\.c&#8203;:1902\]
&nbsp;            ret = NULL;
&nbsp;        
&nbsp;            /\* Pick up the branches\, linking them together\. \*/
&nbsp;     =>     br = regbranch\(&flags\, 1\);
&nbsp;            if \(br == NULL\)
&nbsp;            return\(NULL\);
&nbsp;            if \(\*PL\_regcomp\_parse == '|'\) \{
&nbsp;    Perl\_pregcomp  \[regcomp\.c&#8203;:1415\]
&nbsp;            PL\_regcode\->next\_off = \(\(PL\_seen\_evals > U16\_MAX\) ? U16\_MAX :
PL\_seen\_evals\);
&nbsp;            REGC\(\(U8\)REG\_MAGIC\, \(char\*\) PL\_regcode\+\+\);
&nbsp;            r\->data = 0;
&nbsp;     =>     if \(reg\(0\, &flags\) == NULL\)
&nbsp;            return\(NULL\);
&nbsp;        
&nbsp;            /\* Dig out information for optimizations\. \*/
&nbsp;    Perl\_pmruntime \[op\.c&#8203;:2859\]
&nbsp;            \}
&nbsp;            if \(\(PL\_hints & HINT\_UTF8\) || \(SvUTF8\(pat\) && \!\(PL\_hints &
HINT\_BYTE\)\)\)
&nbsp;                pm\->op\_pmdynflags |= PMdf\_UTF8;
&nbsp;     =>     pm\->op\_pmregexp = CALLREGCOMP\(aTHX\_ p\, p \+ plen\, pm\);
&nbsp;            if \(strEQ\("\\\\s\+"\, pm\->op\_pmregexp\->precomp\)\)
&nbsp;                pm\->op\_pmflags |= PMf\_WHITE;
&nbsp;            op\_free\(expr\);
&nbsp;    unknown\_function \[perly\.y&#8203;:648\]
&nbsp;            |    FUNC1 '\(' expr '\)'
&nbsp;                    \{ $$ = newUNOP\($1\, 0\, $3\); \}
&nbsp;            |    PMFUNC '\(' term '\)'
&nbsp;     =>             \{ $$ = pmruntime\($1\, $3\, Nullop\); \}
&nbsp;            |    PMFUNC '\(' term '\,' term '\)'
&nbsp;                    \{ $$ = pmruntime\($1\, $3\, $5\); \}
&nbsp;            |    WORD
&nbsp;    S\_parse\_body   \[perl\.c&#8203;:1249\]
&nbsp;        
&nbsp;            SETERRNO\(0\,SS$\_NORMAL\);
&nbsp;            PL\_error\_count = 0;
&nbsp;     =>     if \(yyparse\(\) || PL\_error\_count\) \{
&nbsp;            if \(PL\_minus\_c\)
&nbsp;                Perl\_croak\(aTHX\_ "%s had compilation errors\.\\n"\,
PL\_origfilename\);
&nbsp;            else \{
====================================================
Mark Leighton Fisher    Thomson Consumer Electronics
fisherm@&#8203;tce\.com         Indianapolis\, IN\, USA
"Display some adaptability\."  \-\- Doug Shaftoe\, \_Cryptonomicon\_

**Comment by p5pRT** [comment]
### From @vanstyn

I couldn't reproduce this with current bleadperl\.

:Problem\#2&#8203;:
:  As I noted before\, program execution results in&#8203;:
:      foo at \(re\_eval 1\) line 1\.
:      Segmentation fault
:  However\, if you comment out the "use re 'eval'" line\, you then get
:       at \(re\_eval 1\) line 1\.
:      Segmentation fault
:  which seems odd\.  Again\, this oddness does not occur with 5\.6\.0\.
:  Using "no re 'eval'" doesn't make the oddness go away\.

I couldn't reproduce this with current bleadperl\.

:Problem\#3&#8203;:
:  In trying to make a t/op test for problem\#2\, I ran into yet another
:  problem\. The program&#8203;:
:
:      eval \{
:      $\_ = "1";
:      our $foo = qr\{ \(\\d\+\)
:             \(?\{ die "foo" \}\)
:              \(?\{ $undefined \}\) \}x;
:      m\{ \(??\{ $foo \}\) \}x;
:      \};
:
:      print "result is \[$@&#8203;\]\\n";
:
:  Generates&#8203;:
:   Modification of a read\-only value attempted at t/op/regex\-oddness\.t line 11\.

This still happens\, and appears to be related to the 'our'; shortest test
case I could come up with was&#8203;:
&nbsp;  crypt% \./perl \-wle 'eval \{our $x; "x"=~/\(?\{ die "foo" \}\)/\}; print "die&#8203;: \[$@&#8203;\]"'
&nbsp;  Name "main&#8203;::x" used only once&#8203;: possible typo at \-e line 1\.
&nbsp;  Modification of a read\-only value attempted at \-e line 1\.
&nbsp;  crypt% 

:Problem\#4&#8203;:
:
:  If you then add a print before the result\-print line\,
:
:       eval \{
:       $\_ = "1";
:       our $foo = qr\{ \(\\d\+\)
:              \(?\{ die "foo" \}\)
:               \(?\{ $undefined \}\) \}x;
:       m\{ \(??\{ $foo \}\) \}x;
:       \};
:
:       print "before\\n"
:       print "result is \[$@&#8203;\]\\n";
:
:  it then segfaults&#8203;:

I couldn't reproduce the segfault with current bleadperl\.

:Problem\#5&#8203;:
:
:  If you take the first program above and inline the $foo regex directly into
:  the main regex\, as with&#8203;:
:
:    use strict;
:    use re 'eval';
:
:    $\_ = "1";
:
:    m\{ \(??\{ qr\{\(\\d\+\)
:            \(?\{ warn "foo" \}\)
:            \(?\{ warn $undefined \}\)\} \}\) \}x;
:
:  you then get
:      panic&#8203;: unknown regstclass 0 at /tmp/p line 9\.

I couldn't reproduce this with current bleadperl\.

:Problem\#6&#8203;:
:
: Making the program even smaller
:
:     use re 'eval';
:     m\{ \(??\{ qr\{\(\\d\+\)\} \}\) \}x;
:
: and you get a segfault \(gdb dump below\)\.

I couldn't reproduce this with current bleadperl\.

:Problem \#7&#8203;: then\, removing the "use re 'eval'" from that small one\, you
:get the panic again&#8203;:
:
:    jfriedl@&#8203;fummy> \./perl \-we 'm\{ \(??\{ qr\{\(\\d\+\)\} \}\) \}x;'
:    Use of uninitialized value in pattern match \(m//\) at \-e line 1\.
:    panic&#8203;: unknown regstclass 0 at \-e line 1\.

I couldn't reproduce this with current bleadperl\.

Hugo

**Comment by p5pRT** [comment]
### From @chorny

http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=3646
All are fixed except for problem 3&#8203;:
eval \{
 $\_ = "1";
 our $foo = qr\{ \(\\d\+\)
&nbsp;  \(?\{ die "foo" \}\)
&nbsp;  \(?\{ $undefined \}\) \}x;
 m\{ \(??\{ $foo \}\) \}x;
\};

print "result is \[$@&#8203;\]\\n";

Now it segfaults in both 5\.8\.8 and 5\.10\.

\-\- 
Alexandr Ciornii\, http&#8203;://chorny\.net

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Dec 15\, 2008 at 11&#8203;:02&#8203;:43AM \-0800\, Alexandr Ciornii via RT wrote&#8203;:
> http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=3646
> All are fixed except for problem 3&#8203;:
> eval \{
>  $\_ = "1";
>  our $foo = qr\{ \(\\d\+\)
>   \(?\{ die "foo" \}\)
>   \(?\{ $undefined \}\) \}x;
>  m\{ \(??\{ $foo \}\) \}x;
> \};
> 
> print "result is \[$@&#8203;\]\\n";
> 
> Now it segfaults in both 5\.8\.8 and 5\.10\.

A new data point&#8203;:

it segfaults on 5\.8\.9\, 5\.10\.x and bleed\, but all only on threaded builds\.

\-\- 
My Dad used to say 'always fight fire with fire'\, which is probably why
he got thrown out of the fire brigade\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Dec 15 11&#8203;:02&#8203;:41 2008\, chorny wrote&#8203;:
> http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=3646
> All are fixed except for problem 3&#8203;:
> eval \{
>  $\_ = "1";
>  our $foo = qr\{ \(\\d\+\)
>   \(?\{ die "foo" \}\)
>   \(?\{ $undefined \}\) \}x;
>  m\{ \(??\{ $foo \}\) \}x;
> \};
> 
> print "result is \[$@&#8203;\]\\n";
> 
> Now it segfaults in both 5\.8\.8 and 5\.10\.
> 

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
